apps/champions/lib/champions/items.ex

```diff
@@ -12,36 +14,94 @@ defmodule Champions.Items do
-  Returns `{:error, :cant_afford}` if user cannot afford the cost.
-  Returns `{:ok, item: %Item{}, user_currency: %UserCurrency{}}` if succesful.
+  Get all items owned by a user.
   """
-  def level_up(user_id, item_id) do
-    with {:item, {:ok, item}} <- {:item, Items.get_item(item_id)},
-         {:item_owned, true} <- {:item_owned, item.user_id == user_id},
-         [{level_up_currency_id, level_up_cost}] = calculate_level_up_cost(item),
-         {:can_afford, true} <-
-           {:can_afford, Currencies.can_afford(user_id, level_up_currency_id, level_up_cost)} do
-      Items.level_up(item, level_up_currency_id, level_up_cost)
-    else
-      {:item, {:error, :not_found}} -> {:error, :not_found}
-      {:item_owned, false} -> {:error, :not_found}
-      {:can_afford, false} -> {:error, :cant_afford}
-    end
-  end
+  def get_items(user_id), do: Items.get_items(user_id)
 
+  @doc """
+  Equip an item to a unit.
+  """
   def equip_item(user_id, item_id, unit_id) do
     Items.equip_item(user_id, item_id, unit_id)
     get_item(item_id)
   end
 
+  @doc """
+  Unequip an item from a unit.
+  """
   def unequip_item(user_id, item_id) do
     Items.unequip_item(user_id, item_id)
     get_item(item_id)
   end
 
-  defp calculate_level_up_cost(item),
-    do: [{Currencies.get_currency_by_name!("Gold").id, item.level |> Math.pow(2) |> round()}]
+  @doc """
+  Consume a list of items that meet specific rank and character requirements based on the target
+  item in order to increase its tier. Unlike Units, Items don't have a tier. Instead, we create a new
+  one with the improved ItemTemplate.
+
+  Returns `{:ok, item}` or `{:error, reason}`.
+  """
+  def fuse(user_id, consumed_items_ids) do
+    with consumed_items <- Items.get_items_by_ids(consumed_items_ids),
+         {:consumed_items_owned, true} <-
+           {:consumed_items_owned, Enum.all?(consumed_items, &(&1.user_id == user_id))},
+         {:consumed_items_count, true} <-
+           {:consumed_items_count, Enum.count(consumed_items) == Enum.count(consumed_items_ids)},
+         {:consumed_items_valid, {true, new_template}} <-
+           {:consumed_items_valid, meets_fuse_requirements?(consumed_items)},
+         {:can_afford, true} <-
+           {:can_afford, Currencies.can_afford(user_id, new_template.upgrade_costs)} do
+      result =
+        Multi.new()
+        |> Multi.run(:item, fn _, _ -> Items.insert_item(%{user_id: user_id, template_id: new_template.id}) end)
+        |> Multi.run(:deleted_items, fn _, _ -> delete_consumed_items(consumed_items_ids) end)
+        |> Multi.run(:currency_deduction, fn _, _ ->
+          Currencies.substract_currencies(user_id, new_template.upgrade_costs)
+        end)
+        |> Transaction.run()
+
+      case result do
+        {:error, _, _, _} ->
+          {:error, :transaction}
+
+        {:ok, %{item: item}} ->
+          Items.get_item(item.id)
```

Comment:
why not simply returning the item?

Reply:
This is to get the preloads that protobuf messages need